pennylane/transforms/optimization/pattern_matching.py

```diff
@@ -0,0 +1,1594 @@
+
+    for i, node in enumerate(circuit_dag.get_nodes()):
+        # Wires
+        wires.append([])
+        for q in node[1].wires:
+            if q in qubits_conf:
+                wires[i].append(qubits_conf.index(q))
+        if len(node[1].wires) != len(wires[i]):
+            wires[i] = []
+
+        # Control wires
+        control_wires.append([])
+        for q in node[1].control_wires:
+            if q in qubits_conf:
+                control_wires[i].append(qubits_conf.index(q))
+        if len(node[1].control_wires) != len(control_wires[i]):
+            control_wires[i] = []
+
+        # Target wires
+        target_wires.append([])
+        for q in node[1].target_wires:
+            if q in qubits_conf:
+                target_wires[i].append(qubits_conf.index(q))
+        if len(node[1].target_wires) != len(target_wires[i]):
+            target_wires[i] = []
+
+    return wires, target_wires, control_wires
+
+
+def _add_match(match_list, backward_match_list):
+    """
+    Add a match configuration found by pattern matching if it is not already in final list of matches.
+    If the match is already in the final list, the qubit configuration is added to the existing Match.
+    Args:
+        match_list (list): match from the backward part of the algorithm.
+        backward_match_list (list): List of matches found by the algorithm for a given configuration.
+    """
+
+    already_in = False
+
+    for b_match in backward_match_list:
+        for l_match in match_list:
+            if b_match.match == l_match.match:
+                index = match_list.index(l_match)
+                match_list[index].qubit.append(b_match.qubit[0])
+                already_in = True
+
+        if not already_in:
+            match_list.append(b_match)
+
+
+def _compare_qubits(node1, wires1, control1, target1, wires2, control2, target2):
+    """Compare the qubit configurations of two operations. The operations are supposed to be similar up to their
+    qubits configuration.
+    Args:
+        node1 (.CommutationDAGNode): First node.
+        wires1 (list(int)): Wires of the first node.
+        control1 (list(int)): Control wires of the first node.
+        target1 (list(int)): Target wires of the first node.
+        wires2 (list(int)): Wires of the second node.
+        control2 (list(int)): Control wires of the second node.
+        target2 (list(int)): Target wires of the second node.
+    """
+    if control1:
+        if set(control1) == set(control2):
+            if node1.op.name in symmetric_over_all_wires:
+                if set(target1) == set(target2):
+                    return True
+            else:
+                if target1 == target2:
+                    return True
+    else:
+        if node1.op.name in symmetric_over_all_wires:
+            if set(wires1) == set(wires2):
+                return True
+        else:
+            if wires1 == wires2:
+                return True
+    return False
+
+
+class ForwardMatch:
```

Comment:
Looking at the class, I am less convinced this needs to be a class when a function, even a function with nested inner functions, would do.

We would have only one instance of the class at a time. We have fixed inputs and outputs. The entire point of the class is to create outputs, not to store information long term. 

If we need to better group the helper functions together, we could just move everything to its own file.

This is all opinion, so feel free to ignore these comments. This probably isn't going to be a highly trafficked area of code, so we don't need highly optimized design; just something that works.

Reply:
Yes your points are all valid, I went with a class implementation purely to separate everything clearly (for me at least) and also preferred to gather everything in one file. I also agree that this code base will not be changed much and that both implementations have pros and cons.